File: scripts/harness_validation/line_limits.py
"""复用下游文件行数检查器，把报告转换为 Harness validator 错误。"""

from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path

from .context import LINE_LIMIT_CHECKER, ROOT, display_path, fail


def validate_repository_line_limits(
    errors: list[str],
    *,
    warnings: list[str] | None = None,
    root: Path = ROOT,
    checker: Path = LINE_LIMIT_CHECKER,
) -> None:
    """运行唯一检查器；软候选按需返回，硬超限和运行错误始终阻断。"""

    if not checker.is_file():
        fail(errors, f"missing file line-limit checker: {display_path(checker)}")
        return
    try:
        result = subprocess.run(
            [sys.executable, str(checker), "--root", str(root), "--json"],
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            check=False,
        )
    except OSError as error:
        fail(errors, f"cannot execute file line-limit checker: {error}")
        return
    try:
        report = json.loads(result.stdout)
    except json.JSONDecodeError as error:
        fail(errors, f"invalid file line-limit checker JSON: {error}")
        return
    if not isinstance(report, dict):
        fail(errors, "file line-limit checker JSON must be an object")
        return
    for detail in report.get("errors", []):
        fail(errors, f"file line-limit checker error: {detail}")
    for candidate in report.get("reviewCandidates", []):
        if not isinstance(candidate, dict):
            fail(errors, f"invalid file line-limit review candidate: {candidate!r}")
            continue
        if warnings is not None:
            warnings.append(
                f"{candidate.get('profile', 'maintained_text')} file exceeds its "
                f"{candidate.get('threshold')}-line refactor-review threshold: "
                f"{candidate.get('path')} ({candidate.get('lines')} lines, "
                f"hard limit {candidate.get('limit')})"
            )
    for violation in report.get("violations", []):
        if not isinstance(violation, dict):
            fail(errors, f"invalid file line-limit violation: {violation!r}")
            continue
        fail(
            errors,
            f"{violation.get('profile', 'maintained_text')} file exceeds its hard "
            f"{violation.get('limit')}-line limit: {violation.get('path')} "
            f"({violation.get('lines')} lines)",
        )
    if result.returncode not in {0, 1, 2}:
        fail(errors, f"file line-limit checker returned invalid exit code: {result.returncode}")
    elif result.returncode == 0 and not report.get("ok"):
        fail(errors, "file line-limit checker returned success with ok=false")
    elif result.returncode != 0 and report.get("ok"):
        fail(errors, "file line-limit checker returned failure with ok=true")
    elif result.returncode != 0 and not report.get("errors") and not report.get("violations"):
        fail(errors, "file line-limit checker failed without diagnostics")

File: scripts/harness_validation/line_limits.mjs
import path from "node:path";

import { inspectRepository } from "../../.agents/skills/desktop-implement-change/scripts/check_file_line_limits.mjs";
import {
  ROOT,
  decodeMaintainedText,
  fail,
  physicalLineCount,
  trackedFiles,
} from "./core.mjs";

const FRONTEND_SUFFIXES = new Set([
  ".astro", ".cjs", ".css", ".cts", ".html", ".js", ".jsx", ".less", ".mjs", ".mts",
  ".sass", ".scss", ".svelte", ".ts", ".tsx", ".vue",
]);
const GENERATED_NAMES = new Set([
  "Cargo.lock", "package-lock.json", "pnpm-lock.yaml", "yarn.lock",
]);

/** 按工程规则选择建议阈值与硬上限。 */
export function lineLimitProfile(relative) {
  const suffix = path.extname(relative).toLowerCase();
  if (suffix === ".rs") return { name: "Rust", review: 400, hard: 800 };
  if (FRONTEND_SUFFIXES.has(suffix)) return { name: "前端", review: 500, hard: 1000 };
  return { name: "人工维护文本", review: 500, hard: 2000 };
}

function shouldSkip(relative) {
  if (GENERATED_NAMES.has(path.posix.basename(relative))) return true;
  if (relative.startsWith(".git/") || relative.startsWith("release/") || relative.startsWith("node_modules/") || relative.startsWith("target/")) return true;
  return false;
}

/** 检查分层物理行数；建议区间仅在显式发布审查时产生提示。 */
export function validateLineLimits(
  errors,
  warnings,
  { releaseReview = false, files = null, inspect = inspectRepository, root = ROOT } = {},
) {
  if (files !== null) {
    // 单元测试可用精确文件集合验证 profile；仓库级门禁始终复用下游唯一检查器。
    let checked = 0;
    for (const relative of files) {
      if (shouldSkip(relative)) continue;
      const text = decodeMaintainedText(path.join(ROOT, relative));
      if (text === null) continue;
      checked += 1;
      const lines = physicalLineCount(text);
      const profile = lineLimitProfile(relative);
      if (lines > profile.hard) fail(errors, `${profile.name}文件超过 ${profile.hard} 行硬上限: ${relative} (${lines} 行)`);
      else if (releaseReview && lines > profile.review) warnings.push(`${profile.name}文件进入 ${profile.review + 1}-${profile.hard} 行语义复核区间: ${relative} (${lines} 行)`);
    }
    return { checked };
  }
  let report;
  try {
    report = inspect(root);
  } catch (error) {
    fail(errors, `cannot execute file line-limit checker: ${error.message}`);
    return { checked: 0 };
  }
  for (const detail of report.errors ?? []) fail(errors, `file line-limit checker error: ${detail}`);
  for (const candidate of report.reviewCandidates ?? []) {
    if (releaseReview) warnings.push(`${candidate.profile} file exceeds its ${candidate.threshold}-line refactor-review threshold: ${candidate.path} (${candidate.lines} lines, hard limit ${candidate.limit})`);
  }
  for (const violation of report.violations ?? []) {
    fail(errors, `${violation.profile} file exceeds its hard ${violation.limit}-line limit: ${violation.path} (${violation.lines} lines)`);
  }
  if (report.ok && ((report.errors?.length ?? 0) > 0 || (report.violations?.length ?? 0) > 0)) fail(errors, "file line-limit checker returned success with diagnostics");
  if (!report.ok && (report.errors?.length ?? 0) === 0 && (report.violations?.length ?? 0) === 0) fail(errors, "file line-limit checker failed without diagnostics");
  return { checked: report.checkedTextFiles ?? 0 };
}

/** 显式发布审查时报告临时标记，日常校验不产生噪声。 */
export function validateReviewMarkers(warnings, files = null) {
  const candidates = files ?? trackedFiles();
  const reviewSuffixes = new Set([".cjs", ".js", ".mjs", ".ps1", ".rs", ".sh", ".ts", ".tsx"]);
  const commentMarker = /^\s*(?:(?:\/\/)|#).*\b(?:TODO|FIXME|HACK)\b/u;
  for (const relative of candidates) {
    if (shouldSkip(relative)) continue;
    if (!relative.startsWith("scripts/") && !relative.startsWith(".agents/skills/")) continue;
    if (!reviewSuffixes.has(path.extname(relative).toLowerCase())) continue;
    const absolute = path.join(ROOT, relative);
    let text;
    try {
      text = decodeMaintainedText(absolute);
    } catch (error) {
      if (error?.code === "ENOENT") continue;
      throw error;
    }
    if (text === null) continue;
    const lines = text.split(/\r\n|[\n\r]/u);
    lines.forEach((line, index) => {
      if (commentMarker.test(line)) {
        warnings.push(`临时标记需语义复核: ${relative}:${index + 1}`);
      }
    });
  }
}
